src/disco/downloader.js

```diff
@@ -0,0 +1,133 @@
+];
+const DISCO_METADATA_URL = "https://disco.bii.a-star.edu.sg/disco_v3_api/toolkit/getSampleMetadata";
+
+const DISCO_DOI_URL = "https://doi.org/10.5281/zenodo.14159931";
+// add more declrations as needed
+
+const ORGAN_MAPPING = {
+
+  // Add more mappings as needed
+};
+
+const execFile = promisify(callbackExecFile);
+
+/** @implements {IDownloader} */
+export class Downloader {
+  constructor(config) {
+    /** @type {Config} */
+    this.config = config;
+    /** @type {string} */
+    this.cacheDir = getCacheDir(config);
+    /** @type {string} */
+    this.baseUrl = config.get(DISCO_BASE_URL, DEFAULT_DISCO_BASE_URL); 
+    /** @type {string[]} */
+    this.batchFiles = DISCO_BATCH_URLS.map(url => {
+      const batchName = url.match(/batch_\d+\.tar\.gz/)[0];
+      return join(this.cacheDir, batchName);
+    });
+    /** @type {string} */
+    this.extractScriptFile = 'extract_dataset.py';
+    /** @type {string} */
+    this.metadataFile = "disco_metadata.tsv";
+        /** @type {string} */
+    this.metadataFilePath = join(this.cacheDir, this.metadataFile);
+    /** @type {string} */
+    this.extractScriptFilePath = getSrcFilePath(
+      config,
+      'disco',
+      this.extractScriptFile
+    );
+    /** @type {OrganMetadataCollection} */
+    this.organMetadata = undefined;
+  }
+
+  async prepareDownload(datasets) {
+
+    this.organMetadata = await OrganMetadataCollection.load(this.config);
+      await downloadFile(this.metadataFilePath, DISCO_METADATA_URL, {
+        overwrite: this.config.get(FORCE, false)
+      });
+
+    // Download all batch tar files
+    for (let i = 0; i < DISCO_BATCH_URLS.length; i++) {
+      await downloadFile(
+        this.batchFiles[i], 
+        DISCO_BATCH_URLS[i], 
+        { overwrite: this.config.get(FORCE, false) }
+      );
+    }
+
+    for (const dataset of datasets) {
+      dataset.dataset_id = `${DISCO_DOI_URL}#${dataset.id}`;
+      dataset.publication = DISCO_DOI_URL;
+      dataset.publication_title = DISCO_PUBLICATION_NAME;
+      dataset.publication_lead_author = DISCO_PUBLICATION_LEAD_AUTHOR;
+      dataset.consortium_name = 'DISCO';
+      dataset.provider_name = 'DISCO';
+      dataset.provider_uuid = 'your-provider-uuid'; // Update with actual UUID
+      dataset.dataset_link = `${this.baseUrl}/datasets/${dataset.id}`;
+      dataset.dataset_technology = 'OTHER';
+    }
+  }
+
+  async download(dataset) {
+    // Execute Python script to extract the dataset using the mapping file
+    const { stdout } = await execFile('python3', [
+      this.extractScriptFilePath,
+      '--metadata', this.metadataFilePath,
+      '--dataset', dataset.id,
+      '--output', dataset.dataFilePath,
+      // '--data-dir', getCacheDir(this.config),
+      ...this.batchFiles
+      
+    ]);
+
+    // Parse metadata from stdout
+    const cell_count_match = /cell_count:\s*(\d+)\s*\n/i.exec(stdout);
+    dataset.dataset_cell_count = parseInt(cell_count_match?.[1]);
```

Comment:
I know this was how it was done in some of the other portals, but see TS2 portal for a better way: https://github.com/hubmapconsortium/hra-workflows-runner/blob/main/src/ts2/downloader.js#L137-L148